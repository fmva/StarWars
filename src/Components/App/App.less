@import "../../Less/overall.less";

main {
  display: block;
}

.main-menu{
  height: calc(100% - @header-height);
  border: 1px solid #cfd3e2;
  box-shadow: 0 0 6px #a0a2a9;
  display: -webkit-box;
  display: -ms-flexbox;
  display: -webkit-flex;
  display: flex;

  /*height:400px;*/
  @media @phone {
    display: block;
   }
}

.spin-loading{
  position: absolute;
  height: 100%;
  width: 100%;
  top: 0;
  background-color: #0e191c;
  opacity: 0.8;
  z-index: 9999;
}

.spin-loading-inner{
  position: absolute;
  left: 50%;
  top: 50%;
  transform: translate(-50%, -50%);
}

