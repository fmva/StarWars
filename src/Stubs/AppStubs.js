const defaultState = {
  loading: false,
  errorMessage: '',
  isDrawerMenu: false,
  isMobileShowFilter: false,
  filter: [],
  filterAdmitted: [],
  subjectList: [],
  titleSubjectField: '',
  titleSubject: '',
  isPaginationButton: false,
  nextPage: null,
  details: [],
  modalDetails: [],
  searchValue: ''
};

const fetchValue = {
  data: {
    count: 87,
    next: 'https://swapi.co/api/people/?page=2',
    previous: null,
    results: [
      {
        name: 'Luke Skywalker',
        height: '172',
        mass: '77',
        hair_color: 'blond',
        skin_color: 'fair',
        eye_color: 'blue',
        birth_year: '19BBY',
        gender: 'male'
      },
      {
        name: 'C-3PO',
        height: '167',
        mass: '75',
        hair_color: 'n/a',
        skin_color: 'gold',
        eye_color: 'yellow',
        birth_year: '112BBY',
        gender: 'n/a'
      }
    ]
  },
  error: {
    status: false,
    text: ''
  }
};

const fetchFilterValue = {
  data: {
    count: 87,
    next: null,
    previous: null,
    results: [
      {
        name: 'Owen Lars',
        height: '178',
        mass: '120',
        hair_color: 'brown, grey',
        skin_color: 'light',
        eye_color: 'blue',
        birth_year: '52BBY',
        gender: 'male'
      },
      {
        name: 'Beru Whitesun lars',
        height: '165',
        mass: '75',
        hair_color: 'brown',
        skin_color: 'light',
        eye_color: 'blue',
        birth_year: '47BBY',
        gender: 'female'
      },
      {
        name: 'R5-D4',
        height: '97',
        mass: '32',
        hair_color: 'n/a',
        skin_color: 'white, red',
        eye_color: 'red',
        birth_year: 'unknown',
        gender: 'n/a'
      }
    ]
  },
  error: {
    status: false,
    text: ''
  }
};

const getFetchValue = fetchData => {
  let results = fetchData.data.results.map(item => {
    return Object.assign({}, item);
  });
  let data = {},
    error = {};
  for (let prop in fetchData.data) {
    if (prop !== 'results') {
      data[prop] = fetchData.data[prop];
    }
  }
  for (let prop in fetchData.error) {
    error[prop] = fetchData.error[prop];
  }
  data.results = results;

  return {
    data: data,
    error: error
  };
};

const filter = [
  {
    title: 'title hair_color',
    value: '',
    name: 'hair_color'
  },
  {
    title: 'title birth_year',
    value: '',
    name: 'birth_year'
  },
  {
    title: 'title birth_year',
    value: '',
    name: 'eye_color'
  }
];

const getFilter = () => {
  return filter.map(item => {
    return Object.assign({}, item);
  });
};

const detail = [
  {
    title: 'name_title',
    name: 'name'
  },
  {
    title: 'birth_year_title',
    name: 'birth_year'
  },
  {
    title: 'eye_color_title',
    name: 'eye_color'
  }
];

const getDetail = () => {
  return detail.map(item => {
    return Object.assign({}, item);
  });
};

let scrollableDiv = {
  current: {
    scrollTop: 0,
    offsetHeight: 0,
    scrollHeight: 0
  }
};

const getScrollableDiv = () => {
  return { current: Object.assign({}, scrollableDiv.current) };
};

const evt = {
  target: {
    value: ''
  }
};

const getEvt = () => {
  return { target: Object.assign({}, evt.target) };
};

let AppStubs = {
  get defaultState() {
    return Object.assign({}, defaultState);
  },
  get fetchValue() {
    return getFetchValue(fetchValue);
  },
  get fetchFilterValue() {
    return getFetchValue(fetchFilterValue);
  },
  get filter() {
    return getFilter();
  },
  get detail() {
    return getDetail();
  },
  get scrollableDiv() {
    return getScrollableDiv();
  },
  get evt() {
    return getEvt();
  },
  namePage: 'people'
};

Object.freeze(AppStubs);

export default AppStubs;
