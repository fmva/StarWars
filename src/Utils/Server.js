import 'whatwg-fetch';

/**
 * server format
 * @return {Object} - json format
 *
 */
let result = () => ({
  data: null,
  error: {
    status: false,
    text: ''
  }
});

/**
 * get data from server
 * @param {string} url - url address
 * @return {Object} - json from server
 *
 */
let _fetch = async function(url) {
  let data,
    json = Object.assign(result(), {});

  try {
    data = await fetch(url);
    if (data.status !== 200) {
      json.error.status = true;
      json.error.text = 'Status: ' + data.status;
      return json;
    }
    json.data = await data.json();
  } catch (e) {
    json.error.status = true;
    json.error.text = e.toString();
  }
  return json;
};

export default {
  fetch: _fetch
};
