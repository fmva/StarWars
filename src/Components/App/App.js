import React, { Component } from 'react';
import './App.less';
import PropTypes from 'prop-types';
import Server from '../../Utils/Server';
import MenuList from '../../Resources/Menu';
import AppData from '../../Resources/AppData';
import Resources from '../../Resources/Resources';
import Settings from '../../Resources/Settings';
import Header from '../Header/Header';
import SubjectsList from '../SubjectsList/SubjectsList';
import Filter from '../Filter/Filter';
import Details from '../Details/Details';
import Mobile from '../../Utils/Mobile';
import Utils from '../../Utils/Utils';
import { Spin, Modal } from 'antd';

const defaultState = {
  loading: false,
  errorMessage: '',
  isDrawerMenu: false,
  isMobileShowFilter: false,
  filter: [],
  filterAdmitted: [],
  subjectList: [],
  titleSubjectField: '',
  titleSubject: '',
  isPaginationButton: false,
  nextPage: null,
  details: [],
  modalDetails: [],
  searchValue: ''
};

let timeoutPrinting = null;

class App extends Component {
  static propTypes = {
    /** name of page */
    namePage: PropTypes.string
  };

  constructor(props) {
    super(props);
    this.state = defaultState;

    this.showDrawerMenu = this.showDrawerMenu.bind(this);
    this.mobileShowFilter = this.mobileShowFilter.bind(this);
    this.admitFilter = this.admitFilter.bind(this);
    this.resetFilter = this.resetFilter.bind(this);
    this.saveFilterData = this.saveFilterData.bind(this);
    this.search = this.search.bind(this);
    this.setPaginationDiv = this.setPaginationDiv.bind(this);
    this.paginationData = this.paginationData.bind(this);
    this.showDetails = this.showDetails.bind(this);
    this.resetSearchValue = this.resetSearchValue.bind(this);
    this.error = this.error.bind(this);
  }

  async componentDidUpdate(prevProps) {
    if (this.props.namePage !== prevProps.namePage) {
      await this.setState({ isDrawerMenu: false });
      await this.loadData(this.props.namePage);
    }
  }

  async componentDidMount() {
    await this.loadData(this.props.namePage);
  }

  /**
   * fetch data from server
   * @param {String} url - url name
   */
  async fetch(url) {
    const result = await Server.fetch(url);

    if (result.error.status) {
      await this.setState({ errorMessage: result.error.text });
      this.error();
    }
    return result;
  }

  /**
   * create new array filter
   * @param {String} filterObject - object with filter property
   */
  createNewFilter(filterObject) {
    return filterObject.filter.map(item => {
      return Object.assign({}, item);
    });
  }

  /**
   * get data from REST API and set state
   * @param {String} namePage - name of page
   */
  async loadData(namePage) {
    this.setState({ loading: true });
    this.scrollableDiv.current.scrollTop = 0;
    const result = await this.fetch(AppData[namePage].path);

    if (result.error.status) {
      await this.setState(defaultState);
      return;
    }

    const state = Object.assign(defaultState, {
      loading: false,
      filter: this.createNewFilter(AppData[namePage]),
      filterAdmitted: this.createNewFilter(AppData[namePage]),
      subjectList: result.data.results,
      details: AppData[namePage].details,
      titleSubjectField: AppData[namePage].titleSubjectField,
      titleSubject: AppData[namePage].title,
      nextPage: result.data.next
    });

    await this.setState(state);
    await this.setState({
      isPaginationButton: this.isPaginationButtonShow(!!result.data.next)
    });
  }

  /**
   * is show/hide pagination button
   * @param {Boolean} isShow - show/hide param
   */
  isPaginationButtonShow(isShow) {
    const offsetHeight = this.scrollableDiv.current.offsetHeight;
    const scrollHeight = this.scrollableDiv.current.scrollHeight;

    if (!isShow) {
      return false;
    }

    return scrollHeight <= offsetHeight + Settings.getSettings('scrollOffset');
  }

  /**
   * save div with pagination
   * @param {Object} scrollableDiv - scrollable div
   */
  setPaginationDiv(scrollableDiv) {
    this.scrollableDiv = scrollableDiv;
  }

  /**
   * show/hide drawer menu
   */
  showDrawerMenu() {
    this.setState({ isDrawerMenu: !this.state.isDrawerMenu });
  }

  /**
   * show/hide filter fields for mobile version
   */
  async mobileShowFilter() {
    if (!Utils.isEqualAdmittedFilter(this.state.filter, this.state.filterAdmitted)) {
      await this.admitFilter();
    } else {
      this.setState({ isMobileShowFilter: !this.state.isMobileShowFilter });
    }
  }

  /**
   * handler admit button
   */
  async admitFilter() {
    await this.setState({
      loading: true
    });
    const listFiltered = await this.getFilterData('');

    await this.setState({
      filterAdmitted: this.createNewFilter(this.state),
      subjectList: listFiltered.data.results,
      nextPage: listFiltered.data.next,
      isMobileShowFilter: Mobile.mobileCheck()
        ? !this.state.isMobileShowFilter
        : this.state.isMobileShowFilter
    });
    await this.setState({
      isPaginationButton:
        listFiltered.data.results.length > 0 &&
        this.isPaginationButtonShow(!!listFiltered.data.next),
      loading: false
    });
  }

  /**
   * handler reset button
   */
  async resetFilter() {
    const filter = this.createNewFilter(AppData[this.props.namePage]);

    if (
      !this.state.filterAdmitted.some(item => {
        return item.value !== '';
      })
    ) {
      await this.setState({
        filter: filter,
        filterAdmitted: this.createNewFilter(AppData[this.props.namePage])
      });
      return;
    }

    await this.setState({
      loading: true,
      filter: filter
    });

    const listFiltered = await this.getFilterData('');

    this.setState({
      filterAdmitted: this.createNewFilter(this.state),
      subjectList: listFiltered.data.results,
      nextPage: listFiltered.data.next
    });
    await this.setState({
      isPaginationButton:
        listFiltered.data.results.length > 0 &&
        this.isPaginationButtonShow(!!listFiltered.data.next),
      loading: false
    });
  }

  /**
   * filter data
   * @param {Array} arr - array of main data
   * @param {Array} filterDescription - array of filter datd
   */
  filterListArray(arr, filterDescription) {
    return arr.filter(itemSubject => {
      return filterDescription.every(itemFilter => {
        return (
          itemSubject[itemFilter.name] &&
          itemSubject[itemFilter.name].toString().includes(itemFilter.value.trim())
        );
      });
    });
  }

  /**
   * take data from server and filtering their
   * @param {String} nextPage - url for nest request
   */
  async getFilterData(nextPage) {
    let localSubjects = [];
    let localNextPage = nextPage
      ? nextPage
      : this.state.searchValue !== undefined && this.state.searchValue !== null
      ? this.getSearchUrl()
      : AppData[this.props.namePage].path;

    while (localSubjects.length < Settings.getSettings('countSubjectItems') && localNextPage) {
      const result = await this.fetch(localNextPage);
      if (result.error.status) {
        await this.setState({ loading: false });
        return;
      }
      localSubjects = localSubjects.concat(
        this.filterListArray(result.data.results, this.state.filter)
      );
      localNextPage = result.data.next;
    }

    return {
      data: {
        results: localSubjects.slice(0, Settings.getSettings('countSubjectItems')),
        next: localNextPage
      }
    };
  }

  /**
   * handler for changing data in a fields
   * @param {Number} id - id of filter array
   * @param {String} value - value of filter field
   */
  saveFilterData(id, value) {
    const filter = this.createNewFilter(this.state);

    filter[id].value = value;
    this.setState({ filter: filter });
  }

  /**
   * get url with search value
   */
  getSearchUrl() {
    const namePage = this.props.namePage;
    const value = this.state.searchValue;

    return AppData[namePage].path + '?' + Settings.getSettings('search') + '=' + value;
  }

  /**
   * request when user scroll down or click button pagination
   */
  async paginationData() {
    if (!this.state.nextPage) {
      return;
    }

    this.scrollableDiv.current.scrollTop =
      this.scrollableDiv.current.scrollHeight - Settings.getSettings('scrollOffset');
    const isNewList = Utils.isEqualAdmittedFilter(this.state.filter, this.state.filterAdmitted)
      ? false
      : true;
    await this.requestItemData(isNewList, Settings.getSettings('timeoutPagination'));
  }

  /**
   * handler for search field
   * @param {Object} evt - object of field
   */
  async search(evt) {
    const value = evt.target.value;

    await this.setState({ searchValue: value });
    await this.requestItemData(true, Settings.getSettings('timeoutPrinting'));
  }

  /**
   * request data for pagination and search
   * @param {Boolean} isNewList - add list to current values or render new list
   * @param {Number} timeout - timeout ms, default 0
   */
  async requestItemData(isNewList, timeout = 0) {
    const url = isNewList ? this.getSearchUrl() : this.state.nextPage;

    clearTimeout(timeoutPrinting);
    return new Promise(resolve => {
      timeoutPrinting = setTimeout(async () => {
        this.setState({ loading: true });

        const result = await this.getFilterData(url);

        const subjectList = isNewList
          ? result.data.results
          : this.state.subjectList.concat(result.data.results);

        await this.setState({
          filterAdmitted: this.createNewFilter(this.state),
          subjectList: subjectList,
          nextPage: result.data.next
        });
        await this.setState({
          isPaginationButton: this.isPaginationButtonShow(!!result.data.next)
        });
        await this.setState({ loading: false });
        resolve();
      }, timeout);
    });
  }

  /**
   * show details for item
   * @param {Number} index - index of item
   */
  showDetails(index) {
    const subjectList = this.state.subjectList;
    let arrColumns = [];
    let cnt = 0;

    if (index === undefined || index === null) {
      this.setState({ modalDetails: [] });
      return;
    }

    const modalDetails = this.state.details
      .map((item, key) => {
        if (cnt === 0) {
          arrColumns = [];
        }
        if (subjectList[index][item.name]) {
          arrColumns.push({
            title: item.title,
            value: subjectList[index][item.name]
          });
          cnt++;
        }

        if (
          cnt === Settings.getSettings('countDetailColumns') ||
          key === this.state.details.length - 1
        ) {
          cnt = 0;
          return arrColumns;
        }
        return null;
      })
      .filter(item => {
        return item && item.length > 0;
      });

    this.setState({ modalDetails: modalDetails });
  }

  /**
   * reset search value
   */
  async resetSearchValue() {
    await this.setState({ searchValue: '' });
    await this.requestItemData(true, Settings.getSettings('timeoutPrinting'));
  }

  /**
   * error modal window
   */
  error() {
    Modal.error({
      title: Resources.getResource('error_message'),
      content: this.state.errorMessage
    });
  }

  render() {
    return (
      <div className="height-100">
        <Header
          menu={MenuList}
          onClickMenu={this.showDrawerMenu}
          isDrawerMenu={this.state.isDrawerMenu}
        />
        <main className="main-menu main-size">
          <Filter
            mobileShowFilter={this.state.isMobileShowFilter}
            onClickMobileHeader={this.mobileShowFilter}
            filterInputs={this.state.filter}
            filterAdmitted={this.state.filterAdmitted}
            onClickAdmit={this.admitFilter}
            onClickReset={this.resetFilter}
            saveFilterData={this.saveFilterData}
          />
          <SubjectsList
            isPaginationButton={this.state.isPaginationButton}
            setPaginationDiv={this.setPaginationDiv}
            onPagination={this.paginationData}
            onClickItem={this.showDetails}
            title={this.state.titleSubject}
            subjects={this.state.subjectList}
            titleSubjectField={this.state.titleSubjectField}
            onChangeSearch={this.search}
            searchValue={this.state.searchValue}
            onResetSearchValue={this.resetSearchValue}
          />
        </main>

        <div className="spin-loading" style={{ display: this.state.loading ? 'block' : 'none' }}>
          <div className="spin-loading-inner">
            <Spin
              tip={Resources.getResource('loading')}
              spinning={this.state.loading}
              size="large"
            />
          </div>
        </div>

        <Details
          details={this.state.modalDetails}
          onClose={() => {
            this.showDetails(null);
          }}
        />
      </div>
    );
  }
}

export default App;
