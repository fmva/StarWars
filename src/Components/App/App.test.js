import App from './App';
import Server from '../../Utils/Server';
import { shallow } from 'enzyme/build';
import React from 'react';
import AppStubs from '../../Stubs/AppStubs';

describe('App', () => {
  beforeAll(() => {});

  it('Компонент App с входными данными для актеров', async () => {
    Server.fetch = jest.fn();
    Server.fetch.mockReturnValue(AppStubs.fetchValue);

    const wrapper = shallow(<App namePage={AppStubs.namePage} />, {
      disableLifecycleMethods: true
    });
    wrapper.instance().scrollableDiv = AppStubs.scrollableDiv;

    wrapper.update();

    await wrapper.instance().componentDidMount();
    expect(wrapper).toMatchSnapshot();

    expect(wrapper.state('subjectList').length).toBe(2);
    expect(wrapper.state('subjectList')[0].name).toBe('Luke Skywalker');
    expect(wrapper.state('subjectList')[1].name).toBe('C-3PO');

    expect(wrapper.state('filter').length).toBe(3);
    expect(wrapper.state('filter')[0].title).toBe('Hairs');
    expect(wrapper.state('filter')[1].title).toBe('Birth year');
    expect(wrapper.state('filter')[2].title).toBe('Eye color');

    expect(wrapper.state('details').length).toBe(8);
    expect(wrapper.state('details')[0].title).toBe('Name');
    expect(wrapper.state('details')[1].title).toBe('Birth year');
    expect(wrapper.state('details')[2].title).toBe('Eye color');
    expect(wrapper.state('details')[3].title).toBe('Gender');
    expect(wrapper.state('details')[4].title).toBe('Hairs');
    expect(wrapper.state('details')[5].title).toBe('Height (cm)');
    expect(wrapper.state('details')[6].title).toBe('Mass (kg)');
    expect(wrapper.state('details')[7].title).toBe('Skin');

    expect(wrapper.state('nextPage')).toBe('https://swapi.co/api/people/?page=2');
    expect(wrapper.state('titleSubject')).toBe('People');
    expect(wrapper.state('titleSubjectField')).toBe('name');

    Server.fetch.mockReset();
  });

  it(
    'Если пользователь нажал на иконку открыть/закрыть меню (мобильные устройства), ' +
      'тогда необходимо изменить параметр isDrawerMenu',
    async () => {
      const wrapper = shallow(<App namePage={AppStubs.namePage} />, {
        disableLifecycleMethods: true
      });
      wrapper.instance().loadData = jest.fn();

      wrapper.setState(Object.assign({}, AppStubs.defaultState));
      await wrapper.instance().componentDidMount();

      wrapper.find('Header').prop('onClickMenu')();
      expect(wrapper.state('isDrawerMenu')).toBeTruthy();

      wrapper.instance().loadData.mockReset();
    }
  );

  it(
    'Если пользователь нажал на заголовок фильтра (мобильные устройства), ' +
      'тогда необходимо изменить параметр isMobileShowFilter',
    async () => {
      const wrapper = shallow(<App namePage={AppStubs.namePage} />, {
        disableLifecycleMethods: true
      });
      wrapper.instance().loadData = jest.fn();

      wrapper.setState(Object.assign({}, AppStubs.defaultState));
      await wrapper.instance().componentDidMount();

      wrapper.find('Filter').prop('onClickMobileHeader')();
      expect(wrapper.state('isMobileShowFilter')).toBeTruthy();

      wrapper.instance().loadData.mockReset();
    }
  );

  it(
    'Если пользователь нажал на кнопку применить фильтр, ' +
      'тогда необходимо применить фильтр к данным',
    async () => {
      const wrapper = shallow(<App namePage={AppStubs.namePage} />, {
        disableLifecycleMethods: true
      });
      wrapper.instance().scrollableDiv = AppStubs.scrollableDiv;
      wrapper.instance().loadData = jest.fn();

      const filter = AppStubs.filter;
      filter[2].value = 'blue';

      wrapper.setState(
        Object.assign({}, AppStubs.defaultState, {
          subjectList: [],
          filter: filter
        })
      );

      Server.fetch = jest.fn();
      Server.fetch.mockReturnValue(AppStubs.fetchFilterValue);

      await wrapper.instance().componentDidMount();

      await wrapper.find('Filter').prop('onClickAdmit')();

      expect(wrapper.state('subjectList').length).toBe(2);
      expect(wrapper.state('subjectList')[0].name).toBe('Owen Lars');
      expect(wrapper.state('subjectList')[0].eye_color).toBe('blue');
      expect(wrapper.state('subjectList')[1].name).toBe('Beru Whitesun lars');
      expect(wrapper.state('subjectList')[1].eye_color).toBe('blue');

      wrapper.instance().loadData.mockReset();
      Server.fetch.mockReset();
    }
  );

  it(
    'Если пользователь нажал на кнопку отменить фильтр, ' +
      'тогда необходимо вернуть прежние данные',
    async () => {
      const wrapper = shallow(<App namePage={AppStubs.namePage} />, {
        disableLifecycleMethods: true
      });
      wrapper.instance().scrollableDiv = AppStubs.scrollableDiv;
      wrapper.instance().loadData = jest.fn();

      const filter = AppStubs.filter;
      filter[2].value = 'blue';

      wrapper.setState(
        Object.assign({}, AppStubs.defaultState, {
          subjectList: [],
          filter: filter,
          filterAdmitted: filter
        })
      );

      Server.fetch = jest.fn();
      Server.fetch.mockReturnValue(AppStubs.fetchFilterValue);

      await wrapper.instance().componentDidMount();

      await wrapper.find('Filter').prop('onClickReset')();

      expect(wrapper.state('subjectList').length).toBe(3);
      expect(wrapper.state('subjectList')[0].name).toBe('Owen Lars');
      expect(wrapper.state('subjectList')[0].eye_color).toBe('blue');
      expect(wrapper.state('subjectList')[1].name).toBe('Beru Whitesun lars');
      expect(wrapper.state('subjectList')[1].eye_color).toBe('blue');
      expect(wrapper.state('subjectList')[2].name).toBe('R5-D4');
      expect(wrapper.state('subjectList')[2].eye_color).toBe('red');

      wrapper.instance().loadData.mockReset();
      Server.fetch.mockReset();
    }
  );

  it(
    'Если пользователь ввел данные в поисковую строку, ' +
      'тогда необходимо вернуть данные с сервера',
    async () => {
      const wrapper = shallow(<App namePage={AppStubs.namePage} />, {
        disableLifecycleMethods: true
      });
      wrapper.instance().scrollableDiv = AppStubs.scrollableDiv;
      wrapper.instance().loadData = jest.fn();

      const filter = AppStubs.filter;
      const evt = AppStubs.evt;

      wrapper.setState(
        Object.assign({}, AppStubs.defaultState, {
          subjectList: AppStubs.fetchValue.data.results,
          filter: filter
        })
      );

      Server.fetch = jest.fn();
      Server.fetch.mockReturnValue(AppStubs.fetchFilterValue);

      await wrapper.instance().componentDidMount();

      await wrapper.find('SubjectsList').prop('onChangeSearch')(evt);

      expect(wrapper.state('subjectList').length).toBe(3);
      expect(wrapper.state('subjectList')[0].name).toBe('Owen Lars');
      expect(wrapper.state('subjectList')[0].eye_color).toBe('blue');
      expect(wrapper.state('subjectList')[1].name).toBe('Beru Whitesun lars');
      expect(wrapper.state('subjectList')[1].eye_color).toBe('blue');
      expect(wrapper.state('subjectList')[2].name).toBe('R5-D4');
      expect(wrapper.state('subjectList')[2].eye_color).toBe('red');

      wrapper.instance().loadData.mockReset();
      Server.fetch.mockReset();
    }
  );

  it(
    'Если пользователь выполняет пагинацию, ' + 'тогда необходимо добавить данные к текущим данным',
    async () => {
      const wrapper = shallow(<App namePage={AppStubs.namePage} />, {
        disableLifecycleMethods: true
      });
      wrapper.instance().scrollableDiv = AppStubs.scrollableDiv;
      wrapper.instance().loadData = jest.fn();

      const filter = AppStubs.filter;

      wrapper.setState(
        Object.assign({}, AppStubs.defaultState, {
          nextPage: AppStubs.fetchValue.data.next,
          subjectList: AppStubs.fetchValue.data.results,
          filter: filter,
          filterAdmitted: filter
        })
      );

      Server.fetch = jest.fn();
      Server.fetch.mockReturnValue(AppStubs.fetchFilterValue);

      await wrapper.instance().componentDidMount();

      await wrapper.find('SubjectsList').prop('onPagination')();

      expect(wrapper.state('subjectList').length).toBe(5);
      expect(wrapper.state('subjectList')[0].name).toBe('Luke Skywalker');
      expect(wrapper.state('subjectList')[0].eye_color).toBe('blue');
      expect(wrapper.state('subjectList')[1].name).toBe('C-3PO');
      expect(wrapper.state('subjectList')[1].eye_color).toBe('yellow');
      expect(wrapper.state('subjectList')[2].name).toBe('Owen Lars');
      expect(wrapper.state('subjectList')[2].eye_color).toBe('blue');
      expect(wrapper.state('subjectList')[3].name).toBe('Beru Whitesun lars');
      expect(wrapper.state('subjectList')[3].eye_color).toBe('blue');
      expect(wrapper.state('subjectList')[4].name).toBe('R5-D4');
      expect(wrapper.state('subjectList')[4].eye_color).toBe('red');

      wrapper.instance().loadData.mockReset();
      Server.fetch.mockReset();
    }
  );

  it(
    'Если пользователь выполняет пагинацию, и при этом установлен фильтр, ' +
      'тогда необходимо добавить данные к текущим данным с учетом фильтра',
    async () => {
      const wrapper = shallow(<App namePage={AppStubs.namePage} />, {
        disableLifecycleMethods: true
      });
      wrapper.instance().scrollableDiv = AppStubs.scrollableDiv;
      wrapper.instance().loadData = jest.fn();

      const filter = AppStubs.filter;
      filter[2].value = 'blue';

      wrapper.setState(
        Object.assign({}, AppStubs.defaultState, {
          nextPage: AppStubs.fetchValue.data.next,
          subjectList: AppStubs.fetchValue.data.results,
          filter: filter,
          filterAdmitted: filter
        })
      );

      Server.fetch = jest.fn();
      Server.fetch.mockReturnValue(AppStubs.fetchFilterValue);

      await wrapper.instance().componentDidMount();

      await wrapper.find('SubjectsList').prop('onPagination')();

      expect(wrapper.state('subjectList').length).toBe(4);
      expect(wrapper.state('subjectList')[0].name).toBe('Luke Skywalker');
      expect(wrapper.state('subjectList')[0].eye_color).toBe('blue');
      expect(wrapper.state('subjectList')[1].name).toBe('C-3PO');
      expect(wrapper.state('subjectList')[1].eye_color).toBe('yellow');
      expect(wrapper.state('subjectList')[2].name).toBe('Owen Lars');
      expect(wrapper.state('subjectList')[2].eye_color).toBe('blue');
      expect(wrapper.state('subjectList')[3].name).toBe('Beru Whitesun lars');
      expect(wrapper.state('subjectList')[3].eye_color).toBe('blue');

      wrapper.instance().loadData.mockReset();
      Server.fetch.mockReset();
    }
  );

  it(
    'Если пользователь выбрал список для просмотра деталей, ' +
      'тогда необходимо сформировать данные для этого',
    async () => {
      const wrapper = shallow(<App namePage={AppStubs.namePage} />, {
        disableLifecycleMethods: true
      });
      wrapper.instance().loadData = jest.fn();

      wrapper.setState(
        Object.assign({}, AppStubs.defaultState, {
          nextPage: AppStubs.fetchValue.data.next,
          subjectList: AppStubs.fetchValue.data.results,
          details: AppStubs.detail
        })
      );

      await wrapper.instance().componentDidMount();

      await wrapper.find('SubjectsList').prop('onClickItem')(0);

      expect(wrapper.state('modalDetails').length).toBe(2);
      expect(wrapper.state('modalDetails')[0][0].title).toBe('name_title');
      expect(wrapper.state('modalDetails')[0][0].value).toBe('Luke Skywalker');
      expect(wrapper.state('modalDetails')[0][1].title).toBe('birth_year_title');
      expect(wrapper.state('modalDetails')[0][1].value).toBe('19BBY');
      expect(wrapper.state('modalDetails')[1][0].title).toBe('eye_color_title');
      expect(wrapper.state('modalDetails')[1][0].value).toBe('blue');

      wrapper.instance().loadData.mockReset();
      //Server.fetch.mockReset();
    }
  );
});
