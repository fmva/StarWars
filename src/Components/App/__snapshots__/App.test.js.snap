// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App Компонент App с входными данными для актеров 1`] = `
<div
  className="height-100"
>
  <Header
    isDrawerMenu={false}
    menu={
      Array [
        Object {
          "path": "people",
          "title": "People",
        },
        Object {
          "path": "films",
          "title": "Films",
        },
        Object {
          "path": "starships",
          "title": "Starships",
        },
      ]
    }
    onClickMenu={[Function]}
  />
  <main
    className="main-menu main-size"
  >
    <Filter
      filterAdmitted={
        Array [
          Object {
            "name": "hair_color",
            "title": "Hairs",
            "value": "",
          },
          Object {
            "name": "birth_year",
            "title": "Birth year",
            "value": "",
          },
          Object {
            "name": "eye_color",
            "title": "Eye color",
            "value": "",
          },
        ]
      }
      filterInputs={
        Array [
          Object {
            "name": "hair_color",
            "title": "Hairs",
            "value": "",
          },
          Object {
            "name": "birth_year",
            "title": "Birth year",
            "value": "",
          },
          Object {
            "name": "eye_color",
            "title": "Eye color",
            "value": "",
          },
        ]
      }
      mobileShowFilter={false}
      onClickAdmit={[Function]}
      onClickMobileHeader={[Function]}
      onClickReset={[Function]}
      saveFilterData={[Function]}
    />
    <SubjectsList
      isPaginationButton={true}
      onChangeSearch={[Function]}
      onClickItem={[Function]}
      onPagination={[Function]}
      onResetSearchValue={[Function]}
      searchValue=""
      setPaginationDiv={[Function]}
      subjects={
        Array [
          Object {
            "birth_year": "19BBY",
            "eye_color": "blue",
            "gender": "male",
            "hair_color": "blond",
            "height": "172",
            "mass": "77",
            "name": "Luke Skywalker",
            "skin_color": "fair",
          },
          Object {
            "birth_year": "112BBY",
            "eye_color": "yellow",
            "gender": "n/a",
            "hair_color": "n/a",
            "height": "167",
            "mass": "75",
            "name": "C-3PO",
            "skin_color": "gold",
          },
        ]
      }
      title="People"
      titleSubjectField="name"
    />
  </main>
  <div
    className="spin-loading"
    style={
      Object {
        "display": "none",
      }
    }
  >
    <div
      className="spin-loading-inner"
    >
      <Spin
        size="large"
        spinning={false}
        tip="Loading..."
        wrapperClassName=""
      />
    </div>
  </div>
  <Details
    details={Array []}
    onClose={[Function]}
  />
</div>
`;
